### DISCUSSION TITLE
Best practices for compilation OOM

### DISCUSSION THREAD

**Discussion by OhadRubin** [jax-ml/jax#12502] [Q&A]
Hey,
What are some best for writing jax code that doesn't OOM during compilation?
If understand correctly it uses scans, but I haven't found any specific information related to this.
Thanks!

**Comment by jakevdp** [comment]
Hard to say without more information, but one thing to keep in mind is that JAX compilation unrolls python control flow. So something like
```python
@jit
def f(x):
  for i in range(1000000):
    x += 1
  return x
```
would result in 1,000,000 `add` statements in the IR. So one best practice is don't use large Python `for` loops (see [Sharp Bits: Control Flow](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#control-flow) for more on this).

**Reply by OhadRubin** [reply]
Ok, awesome.
Regarding more information: I am experiencing OOM in compilation after increasing the batch size, my implementation is using jnp.vectorize to handle the multiple dimensions, is that not advisable?

**Reply by jakevdp** [reply]
If `jnp.vectorize` is causing issues, are you certain the OOM is at compilation time and not at runtime? It's possible with `vectorize` to inadvertently create very large arrays given small inputs that might OOM at runtime.

**Reply by OhadRubin** [reply]
It happens during jitting, so I assumed that means compilation.
Using vmap would reduce the memory consumption compared to jnp.vectorize?

**Reply by jakevdp** [reply]
No, `vectorize` is just a wrapper around `vmap` (or multiple `vmap`s)

If you could edit your question with a minimal reproducible example of the issue, it may be possible to give more pointed advice.
